### PR TITLE
Initial implementation from the copy of Embulk v0.10.2

### PR DESCRIPTION
**Pull request by dmikurube**
Reconstructed Embulk's config processor :
* Based on the original `org.embulk.config` implementation.
* With removing any Guice-based mechanisms.
* With a lot of refactoring.
* With more comments and Javadocs.

### CODE REVIEW

**Comment by dmikurube** [comment]
@shroman Can you have a look when you have time?

I know this PR is so big... -- I'm okay to try splitting it if you want. But as almost all of them are mutually related, I guess it can be split into just 2-3 still big PRs. Anyways, not in a much hurry.

See `Task`'s Javadoc and `src/legacyTest` for its rough usage. As a plugin developer's perspective, I tried to keep its usage almost the same with `embulk-core`'s one. Developers just need to build `ConfigMapper` and `TaskMapper` instead of calling `ConfigSource#loadConfig` and `TaskSource#loadTask`, respectively.

I also know it has too few tests. I'll try to add more tests later, but I believe that minimum required use-cases are covered by `legacyTest`, which uses this library in the actual context of Embulk v0.10.2.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,57 @@
...
+      # For example: signing.secretKeyRingFile=/home/you/.gnupg/1234567890ABCDEF1234567890ABCDEF12345678.secring.gpg
+      signing.secretKeyRingFile=<path to your OpenPGP secret key ring file>
+      ```
+
+### Release
+
+1. Create a detached `HEAD` from the `master` branch.
+    * ```
+      git checkout master
+      git checkout --detach
```
Comment (shroman) [inline README.md:35]: Just curious. Is it really necessary to detach? Having all this in `master` branch doesn't seem to hurt.

File: src/main/java/org/embulk/util/config/Version.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    static com.fasterxml.jackson.core.Version parseVersionForTesting(final String version) {
+        return parseVersion(version);
+    }
+
+    private static com.fasterxml.jackson.core.Version getVersion() {
+        try {
+            return parseVersion(getImplementationVersion());
+        } catch (final Throwable any) {  // It must not fail.
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/Version.java:41]: Not sure about the intention of this comment, because it can fail if `getImplementationVersion` returns `null`.

File: src/main/java/org/embulk/util/config/DataSourceSerializer.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    public void serialize(
+            final T value,
+            final JsonGenerator jsonGenerator,
+            final SerializerProvider serializerProvider)
+            throws IOException {
+        if (value == null) {
+            throw new JsonGenerationException(new NullPointerException(
+                    "DataSourceSerializer#serialize accepts only non-null value."));
+        }
+        final String valueJsonStringified = Compat.toJson(value);  // TODO: DataSource#toJson
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/DataSourceSerializer.java:44]: Is it still `TODO`?

File: src/main/java/org/embulk/util/config/ConfigMapper.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * Maps {@code org.embulk.config.ConfigSource} into a task-defining interface that inherits {@link Task}.
+     *
+     * @param <T>  the task-defining interface
+     * @param config  {@code org.embulk.config.ConfigSource} to map from
+     * @param taskType  {@link java.lang.Class} of the task-defining interface
+     * @return a mapped task instance
+     */
+    public <T extends Task> T map(final ConfigSource config, final Class<T> taskType) {
+        final ObjectNode objectNode = Compat.rebuildObjectNode(config);
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/ConfigMapper.java:66]: `Compat#callGetObjectNodeAndRebuildIfAvailable` can throw a bunch of Runtime exceptions. How about catching them and handle explicitly as `ConfigException` too?

Comment (shroman) [reply src/main/java/org/embulk/util/config/ConfigMapper.java:66]: I think it's good to align with `this.objectMapper.readValue(objectNode.traverse(), taskType);` since basically it's a part of the same thing -- getting a `Task`.

File: src/main/java/org/embulk/util/config/TaskValidationException.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import javax.validation.ConstraintViolation;
+import javax.validation.ConstraintViolationException;
+import org.embulk.config.ConfigException;
+
+/**
+ * {@link java.lang.Exception} that represents violation(s) with constraints of {@link javax.validation.Validator}.
+ */
+public class TaskValidationException extends ConfigException {
+    <T> TaskValidationException(final Set<ConstraintViolation<T>> violations) {
+        super("Configuration violates constaints validated in task definition.",
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/TaskValidationException.java:29]: `constaints` -> `constraints`

File: src/main/java/org/embulk/util/config/Tasks.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Utility methods for the tweaks on {@link org.embulk.util.config.Task}s.
+ */
+final class Tasks {
+    private Tasks() {
+        // No instantiation.
+    }
+
+    static Optional<String> getFieldNameFromGetter(final String methodName) {
+        if (methodName.startsWith("get")) {
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/Tasks.java:31]: Potential NPE?

File: src/main/java/org/embulk/util/config/Tasks.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            return Optional.of(methodName.substring(3));
+        }
+        return Optional.empty();
+    }
+
+    static Optional<String> getFieldNameFromSetter(final String methodName) {
+        if (methodName.startsWith("set")) {
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/Tasks.java:38]: Potential NPE?

File: src/main/java/org/embulk/util/config/Version.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            version = versionAndSnapshotInfo[0];
+            snapshotInfo = null;
+        } else if (versionAndSnapshotInfo.length == 2) {
+            version = versionAndSnapshotInfo[0];
+            snapshotInfo = versionAndSnapshotInfo[1];
+        } else {
+            throw new IllegalArgumentException("Empty version string.");
+        }
+
+        final String[] versionSplit = version.split("\\.");
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/Version.java:60]: All the subsequent use of `versionSplit` better to trim() to avoid NFE

File: src/main/java/org/embulk/util/config/Version.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        if (selfJarPathString == null || selfJarPathString.isEmpty()) {
+            throw new Exception();
+        }
+
+        final Manifest manifest;
+        try (final JarFile selfJarFile = new JarFile(selfJarPathString)) {
+            manifest = selfJarFile.getManifest();
+        }
+        if (manifest == null) {
+            throw new Exception();
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/Version.java:125]: Some message would be useful.

File: src/main/java/org/embulk/util/config/Version.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        final Attributes mainAttributes = manifest.getMainAttributes();
+        final String implementationVersion = mainAttributes.getValue(Attributes.Name.IMPLEMENTATION_VERSION);
+        if (implementationVersion == null) {
+            throw new Exception();
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/Version.java:131]: ditto

File: src/main/java/org/embulk/util/config/Version.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+            // Pass-through.
+        }
+
+        return null;
+    }
+
+    private static String getManifestImplementationVersion() throws Exception {
+        final URL selfJarUrl = Version.class.getProtectionDomain().getCodeSource().getLocation();
+        if (selfJarUrl == null || !selfJarUrl.getProtocol().equals("file")) {
+            throw new Exception();
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/Version.java:112]: Some message would be useful.

File: src/main/java/org/embulk/util/config/Version.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        final String selfJarPathString = selfJarUrl.getPath();
+        if (selfJarPathString == null || selfJarPathString.isEmpty()) {
+            throw new Exception();
```
Comment (shroman) [inline src/main/java/org/embulk/util/config/Version.java:117]: Some message would be useful.

File: src/main/java/org/embulk/util/config/Version.java (hunk 8 above)
Comment (dmikurube) [reply src/main/java/org/embulk/util/config/Version.java:60]: I intentionally skipped such good error handling here for the same reason above.

* In case any unexpected errors, it should go `unknownVersion` without any complicated error handling and recovery.
* If such `versionSplit` is needed for a version (specified in JAR Manifest from `build.gradle`), that should be caught by `TestVersion.java` before releasing.

File: src/main/java/org/embulk/util/config/Version.java (hunk 11 above)
Comment (dmikurube) [reply src/main/java/org/embulk/util/config/Version.java:112]: For the same reason above. Even if it throws any Exception, that would not be used anywhere, and just `unknownVersion` is set.

File: src/main/java/org/embulk/util/config/Version.java (hunk 12 above)
Comment (dmikurube) [reply src/main/java/org/embulk/util/config/Version.java:117]: ditto.

File: src/main/java/org/embulk/util/config/Version.java (hunk 9 above)
Comment (dmikurube) [reply src/main/java/org/embulk/util/config/Version.java:125]: ditto.

File: src/main/java/org/embulk/util/config/Version.java (hunk 10 above)
Comment (dmikurube) [reply src/main/java/org/embulk/util/config/Version.java:131]: ditto.

File: src/main/java/org/embulk/util/config/Tasks.java (hunk 6 above)
Comment (dmikurube) [reply src/main/java/org/embulk/util/config/Tasks.java]: Addressed in: c1817238c88eb452ff8dfb6da59d6f09936df65a

File: src/main/java/org/embulk/util/config/Tasks.java (hunk 7 above)
Comment (dmikurube) [reply src/main/java/org/embulk/util/config/Tasks.java]: Addressed in: c1817238c88eb452ff8dfb6da59d6f09936df65a

File: src/main/java/org/embulk/util/config/TaskValidationException.java (hunk 5 above)
Comment (dmikurube) [reply src/main/java/org/embulk/util/config/TaskValidationException.java]: Addressed in: 36dc6c08955ec79c31751b6d4b1a81d7bc2f193e

File: src/main/java/org/embulk/util/config/Version.java (hunk 8 above)
Comment (shroman) [reply src/main/java/org/embulk/util/config/Version.java:60]: got it

File: src/main/java/org/embulk/util/config/DataSourceSerializer.java (hunk 3 above)
Comment (shroman) [reply src/main/java/org/embulk/util/config/DataSourceSerializer.java]: Got it.

File: README.md (hunk 1 above)
Comment (shroman) [reply README.md:35]: Sure, I don't mind. Just new for me.

File: src/main/java/org/embulk/util/config/Version.java (hunk 2 above)
Comment (shroman) [reply src/main/java/org/embulk/util/config/Version.java:41]: Looks okay to me. Thanks for explaining.

**Comment by dmikurube** [comment]
Thanks! Added some more comments in 790d63b68d8fd6e804de61da554de7a713e18717. Then merging it.
